rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,383 @@
+uploader --> objsvc PUT /objects/abcd123  {
+                                            origin: '10.10.10.10',
+                                            contentType: 'application/json',
+                                            contentLength: 1234,
+                                            contentSha256: 'abcdef1234',
+                                            transferLength: 1234,
+                                            transferSha256: 'abcdef1234',
+                                            contentEncdoing: 
+                                            expiration: new Date(),
+                                            parts: [
+                                              sha256: 'abcd1234',
+                                              size: 1024,
+                                            }, {
+                                              sha256: 'abcd1234',
+                                              size: 1024,
+                                            }],
+                                          }
+objsvc --> s3 POST /object-abcd123?uploads
+objsvc <-- s3 200 UploadId=u-123
+uploader <-- objsvc 200 {requests: [{
+                          url: 'http://s3.com/object-abcd123?partNumber=1&uploadId=u-123',
+                          headers: {authorization: 'signature-request-1'},
+                          method: 'PUT'
+                        }, {
+                          url: 'http://s3.com/object-abcd123?partNumber=2&uploadId=u-123',
+                          headers: {authorization: 'signature-request-1'},
+                          method: 'PUT'
+                        }
+                       ]}
+uploader --> s3 PUT /object-abcd123?partNumber=1&uploadId=u-123'
+uploader <-- s3 200 ETag {etag: 123}
+uploader --> s3 PUT /object-abcd123?partNumber=2&uploadId=u-123'
+uploader <-- s3 200 ETag {etag: 456}
+uploader --> objsvc PATCH /objects/abcd123 {etags: [123, 456]}
+objsvc --> s3 POST /object-abcd123?uploadId=u-123 {etags: [123, 456]}
+objsvc <-- s3 200 ETag {etag: 123456}
+uploader <-- objsvc 200
+```
+
+### Part Size
+```
+GET /upload-info
+GET /upload-info/:origin
+```
+
+Multipart uploads in various backing services have differing requirements related
+to the size of each individual part.  This endpoint will return information about
+part sizes for a given origin.  In the case that the origin is not specified, the
+origin of the requester will be used.
+
+The response of this method for an S3 origin will be:
+
+```
+[
+  {
+    type: 'single-part',
+    max_size: 5 * 1024 * 1024,
+    min_size: 0,
+    max_part_count: 1,
+    min_part_count: 1,
+    min_part_size: 5 * 1024 * 1024,
+    max_part_size: 5 * 1024 * 1024,
+    part_size_multiple: 1,
+  },
+  {
+    type: 'multipart',
+    max_size: 5 * 1024 * 1024 * 1024,
+    min_size: 0,
+    max_part_count: 10000,
+    min_part_count: 1,
+    min_part_size: 5 * 1024 * 1024,
+    max_part_size: 5 * 1024 * 1024,
+    part_size_multiple: 1,
+  },
+]
```

Comment:
As a way to mitigate the overhead, we could also find out what the minimum single-part upload size supported by all providers is. And state that for files smaller than that, single part upload always works. Newest azure API supports up-to 256MB if uploading an entire blob in one request, I think parts are now 100MB (just checked the latest API).

So we could say that calling `/upload-slicing?origin=...&size=...` is not necessary if uploading files less than 256 MB as single-part upload.

This way uploaders only need to do the pre-flight `/upload-slicing?origin=...&size=...` call if the file is big. And big files already take a long time to upload, so paying a small overhead is no big deal. Besides I'm more willing to pay for overhead if the file is big, because the potential performance benefits is bigger.

Slicing a small object is always uninteresting. Slicing a big object might lead to performance wins because:
  1.  part sizes can match internal block size used in the backend (s3, gcs, ...),
  2.  parts can be uploaded concurrently,
  3. if a connection breaks retries are faster because parts are smaller.

These things matter more for big objects, than they do for small.